### PR TITLE
Benchmarks: Add Feature - Extend the device manager utility to support more functions.

### PR DESCRIPTION
**Pull request by guoshzhao**
**Description**
Rename `nvidia_helper` utility as `device_manager` module and support more functions:
```
device_manager.get_device_count()
device_manager.get_device_utilization(idx)
device_manager.get_device_temperature(idx)
device_manager.get_device_power_limit(idx)
device_manager.get_device_memory(idx)
device_manager.get_device_row_remapped_info(idx)
device_manager.get_device_ecc_error(idx)
```

### CODE REVIEW

File: superbench/common/utils/nvidia_helper.py [hunk 1]
```diff
@@ -5,18 +5,180 @@
...
+        dbecc (int): the count of double bit ecc error, None means failed to get the data.
+    """
+    output = run_command('nvidia-smi dmon -i {} -s e -c 1'.format(idx))
+    if output.returncode == 0:
+        content = output.stdout.splitlines()
+        if len(content) == 3:
+            ecc = list(filter(None, content[2].split(' ')))
+            if len(ecc) == 4:
+                sbecc = int(ecc[1])
+                dbecc = int(ecc[2])
```
Comment (abuccts) [inline superbench/common/utils/nvidia_helper.py:181]: what if they are 'N/A' on geforce cards?

File: superbench/common/utils/nvidia_helper.py [hunk 2]
```diff
@@ -5,18 +5,180 @@
...
+        Temperature
+            GPU Current Temp                  : 36 C
+
+    Args:
+        idx (int): device index.
+
+    Return:
+        remapped_metrics (dict): the row remapped information, None means failed to get the data.
+    """
+    output = run_command('nvidia-smi -i {} -q'.format(idx))
```
Comment (abuccts) [inline superbench/common/utils/nvidia_helper.py:135]: why not use `nvmlDeviceGetRemappedRows`?

File: superbench/common/utils/nvidia_helper.py (hunk 1 above)
Comment (abuccts) [reply superbench/common/utils/nvidia_helper.py:181]: use `nvmlDeviceGetDetailedEccErrors`?

File: superbench/common/utils/nvidia_helper.py (hunk 2 above)
Comment (guoshzhao) [reply superbench/common/utils/nvidia_helper.py:135]: `nvmlDeviceGetRemappedRows` doesn't have python API support.

File: superbench/common/utils/nvidia_helper.py (hunk 1 above)
Comment (guoshzhao) [reply superbench/common/utils/nvidia_helper.py:181]: `nvmlDeviceGetDetailedEccErrors` is labelled as deprecated:
https://github.com/fbcotter/py3nvml/blob/0aa41c0940c7d971381fddb8382266a819be9f75/py3nvml/py3nvml.py#L3437

> 
>      * @deprecated   This API supports only a fixed set of ECC error locations
>      *               On different GPU architectures different locations are supported
>      *               See \ref nvmlDeviceGetMemoryErrorCounter

I will use `nvmlDeviceGetMemoryErrorCounter` instead, which use correctable for single bit and uncorrectable for double bit.

File: superbench/common/utils/nvidia_helper.py [hunk 3]
```diff
@@ -5,18 +5,185 @@
...
+    corrected_ecc = 0
+    uncorrected_ecc = 0
+    for location_idx in range(nvml.NVML_MEMORY_LOCATION_COUNT):
+        try:
+            count = nvml.nvmlDeviceGetMemoryErrorCounter(
+                handle, nvml.NVML_MEMORY_ERROR_TYPE_CORRECTED, nvml.NVML_VOLATILE_ECC, location_idx
+            )
+            corrected_ecc += count
+        except nvml.NVMLError:
+            pass
```
Comment (cp5555) [inline superbench/common/utils/nvidia_helper.py:179]: If it doesn't work, we cannot know it.

File: superbench/common/utils/nvidia_helper.py [hunk 4]
```diff
@@ -5,18 +5,185 @@
...
+
+        try:
+            count = nvml.nvmlDeviceGetMemoryErrorCounter(
+                handle, nvml.NVML_MEMORY_ERROR_TYPE_UNCORRECTED, nvml.NVML_VOLATILE_ECC, location_idx
+            )
+            uncorrected_ecc += count
+        except nvml.NVMLError:
+            pass
```
Comment (cp5555) [inline superbench/common/utils/nvidia_helper.py:187]: The same question for the `pass`. It look like we need to know the error.

File: superbench/common/utils/nvidia_helper.py (hunk 3 above)
Comment (cp5555) [inline superbench/common/utils/nvidia_helper.py:179]: If it doesn't work, we cannot know it.

Comment (cp5555) [reply superbench/common/utils/nvidia_helper.py:179]: Shall we implement it as get_device_memory?

File: superbench/common/utils/nvidia_helper.py [hunk 5]
```diff
@@ -5,18 +5,185 @@
 
 import py3nvml.py3nvml as nvml
 
+from superbench.common.utils import logger
+from superbench.common.utils import run_command
+
+nvml.nvmlInit()
```
Comment (cp5555) [inline superbench/common/utils/nvidia_helper.py:11]: It's a little bit weird to init it globally.

Comment (cp5555) [reply superbench/common/utils/nvidia_helper.py:11]: It's quite like we should provide a class on this. It can init nvml, handle , and util.

Comment (cp5555) [reply superbench/common/utils/nvidia_helper.py:11]: We can use this class as logger.

File: superbench/common/utils/nvidia_helper.py (hunk 3 above)
Comment (guoshzhao) [reply superbench/common/utils/nvidia_helper.py:179]: There have 8 types of memory location, for the nvml API design, the ecc counter is the total ecc error occurred in all the location:
```
class nvmlMemoryLocation_t(Enum):
    NVML_MEMORY_LOCATION_L1_CACHE = 0
    NVML_MEMORY_LOCATION_L2_CACHE = 1
    NVML_MEMORY_LOCATION_DRAM = 2
    NVML_MEMORY_LOCATION_DEVICE_MEMORY = 2
    NVML_MEMORY_LOCATION_REGISTER_FILE = 3
    NVML_MEMORY_LOCATION_TEXTURE_MEMORY = 4
    NVML_MEMORY_LOCATION_TEXTURE_SHM = 5
    NVML_MEMORY_LOCATION_CBU = 6
    NVML_MEMORY_LOCATION_SRAM = 7
    NVML_MEMORY_LOCATION_COUNT = 8
```
Different arch supports different locations, for example, A100 has `NVML_MEMORY_LOCATION_SRAM` supported but p100 doesn't. P100 has `NVML_MEMORY_LOCATION_TEXTURE_SHM` supported but A100 doesn't.

The error raised here is `NVMLError_NotSupported` for current device. The current implementation can ignore the device type and loop all the location to aggregate the ecc error.

Your suggestion is naming it as `get_device_memory`? Do you mean we can only get the ecc count from `NVML_MEMORY_LOCATION_DEVICE_MEMORY` ?

Comment (guoshzhao) [reply superbench/common/utils/nvidia_helper.py]: Currently nvml.NVMLError uses `pass` and other type of exceptions will log error.

File: superbench/common/utils/nvidia_helper.py (hunk 5 above)
Comment (guoshzhao) [reply superbench/common/utils/nvidia_helper.py]: Done, and abstract the utility as a class.

**Review by cp5555 — APPROVED** [review]

**Review by abuccts — APPROVED** [review]

**Pull request merged**